Elsevier Pure implementation

bitmotivator opened the pull request:
This pull request implements parts of the [Elsevier Pure API](https://api.elsevierpure.com/ws/api/rapidoc.html) for the purposes of fetching projects and people in DAMAP.

Implementing the Pure API will allow several universities using Pure to integrate DAMAP without additional work required.

This initial implementation only reads data exports from Pure, a full implementation including authentication will follow later.

## TODO

- [x] Finish the project integration
- [x] Make the project integration configurable
- [x] Test if the Pure integration works when configured as a replacement for the mock services
- [x] Document how to set up the Pure integration using environment variables

lpandath commented:
@bitmotivator This PR does introduce breaking changes for institutional implementations. 

Our TU Graz PersonService currently implements:
`read(String id, MultivaluedMap<String, String> queryParams)`

But this method now throws RuntimeException by default. All institutional implementations need to update to the new signatures:
`read(String id)` 

Would it be possible to document this as a breaking change and include a migration guide? It would really help to understand how to update the existing implementations.

bitmotivator commented:
@lpandath thank you for your review!

Maybe I missed something, but the way it works in Java is that the [`default` implementation of an interface method is not used when you override it](https://docs.oracle.com/javase/tutorial/java/IandI/defaultmethods.html#extending). The default implementation for `read(String id)` calls `read(String id, MultivaluedMap<String,String> queryParams)`, which is your overridden method.

So, unless I made a mistake in the code, you should be able to continue implementing `read(String id, MultivaluedMap<String,String> queryParams)` and it should work the same. In addition, I will test this with the mock implementation before marking it ready for your review.

lpandath commented on src/main/java/org/damap/base/integration/pure/PureAPI.java:
```diff
@@ -0,0 +1,87 @@
+    long pageSize = 10;
+    boolean finished = false;
+    while (!finished) {
+      PureAPIPaginatedProjectsResponse persons = listAllProjects(pageSize, offset);
```
Variable named persons but contains projects?!

lpandath commented on src/main/java/org/damap/base/integration/pure/PureAPI.java:
```diff
@@ -0,0 +1,87 @@
+    while (!finished) {
+      PureAPIPaginatedProjectsResponse persons = listAllProjects(pageSize, offset);
+      items.addAll(persons.items);
+      if (persons.count <= persons.pageInformation.offset + persons.pageInformation.size) {
```
The offset is set to the current offset instead of incrementing, which will cause an infinite loop no? 
Would it be better:
`offset += pageSize;` or `offset = persons.pageInformation.offset + persons.pageInformation.size;`

lpandath committed:
Add: test demonstrating pagination bug

- infinite loop when fetching >10 projects/persons
- offset = pageInformation.offset doesn't advance offset
- test shows 11 API calls with offset stuck at 0

lpandath reviewed (commented):
Very well structured and clean interfaces! 
The bug would cause problems with large Pure datasets.

bitmotivator committed:
Fixed Pure API pagination bug

bitmotivator committed:
Refactoring to allow for configurable project services

bitmotivator committed:
Refactoring to allow for configurable project services

bitmotivator committed:
Experimental Elsevier Pure integration

bitmotivator committed:
Experimental Elsevier Pure integration

ValentinFutterer commented on src/main/java/org/damap/base/integration/pure/PureAPI.java:
```diff
@@ -0,0 +1,87 @@
+   *
+   * @return a list of all persons in the Pure database.
+   */
+  default List<PureAPIPerson> listAllPersons() {
+    List<PureAPIPerson> items = new ArrayList<>();
+    long offset = 0;
+    long pageSize = 10;
+    boolean finished = false;
+    while (!finished) {
+      PureAPIPaginatedPersonsResponse persons = listAllPersons(pageSize, offset);
+      items.addAll(persons.items);
+      offset = persons.pageInformation.offset + persons.pageInformation.size;
+      if (persons.count <= offset) {
+        finished = true;
+      }
+    }
+    return items;
```
I am not quite sure how this works:
What values do `persons.pageInformation.offset` and `persons.pageInformation.size` return?
If you call `listAllPersons` with offset = 0 and pageSize = 10, will they return 0 and 10 respectively? If yes, then `finsihed` will always be set to true.

bitmotivator replied:
Unfortunately, I have no way to test this, but from the API docs I have to surmise that `persons.pageInformation.offset` will always be the offset passed in, while `persons.pageInformation.size` will either be the number of records currently returned, or the size passed in. However, `persons.count`, as far as I understand it, will always be the total number of records on the server, so `finished` should only be `true` once the `offset` variable reaches that number. @lpandath may be able to confirm if this works properly after testing.

lpandath replied:
@bitmotivator  yes, i can test this that when integration in place.

lpandath replied:
@bitmotivator 
https://github.com/damap-org/damap-backend/pull/405

lpandath committed:
chore: replace PURE API key and update doc

lpandath committed:
add: update PURE env

lpandath committed:
test: retain pure test profile for file-based mock testing

lpandath committed:
chore: add docker compose config PURE integration testing

lpandath committed:
test: add pagination behaviour test

lpandath committed:
spoltess: format  files

lpandath committed:
feat: Remove TU Graz specifics from PURE integration and add mock service

lpandath committed:
Refactor Pure: API integration tests with WireMock and JSON data

bitmotivator committed:
Review feedback

bitmotivator committed:
Spotless

bitmotivator committed:
Removed unnecessary extra line

lpandath committed:
update: wiremock dependency

ValentinFutterer commented on src/main/java/org/damap/base/integration/pure/PureProjectService.java:
```diff
@@ -0,0 +1,137 @@
+    return new ProjectSupplementDO();
+  }
+
+  @Override
+  public ResultList<ProjectDO> getRecommended(Search search) {
+    return this.search(search);
+  }
```
I think that this method here is broken, since the frontend passes an empty search by default - so it will always return all projects (or none, depending on how PURE is implemented). You need some custom code here that searches for all projects that includes the callers PURE ID.

lpandath committed:
Fix WireMock port conflicts and Elsevier Pure integration tests

lpandath committed:
clean: remove placeholder URL from config

ValentinFutterer committed:
Fix a bug that prevents the PURE service from being loaded by CDI

ValentinFutterer committed:
Fix tests by removing bad mocks and general refactoring

ValentinFutterer committed:
execute mvn spotless:apply

ValentinFutterer committed:
Fix get recommended endpoint projects

ValentinFutterer reviewed (approved):
After much heartache, I think we can finally merge 🚀
